packages/@vuepress/plugin-public-files/index.js

```diff
@@ -0,0 +1,34 @@
+const {
+  fs: { existsSync },
+  path: { resolve }
+} = require('@vuepress/shared-utils')
+const CopyPlugin = require('copy-webpack-plugin')
+const mergeable = require('vuepress-mergeable')
+
+module.exports = mergeable((patterns, context) => ({
+  name: `@vuepress/plugin-public-files`,
+
+  chainWebpack (config) {
+    if (!Array.isArray(patterns)) patterns = [patterns]
+
+    config
+      .plugin('copy')
+      .use(CopyPlugin, [patterns.map((pattern) => {
+        pattern = typeof pattern === 'string'
+          ? { from: pattern }
+          : { ...pattern }
```

Comment:
This doesn't make sense.  
You first convert everything into an array and then checking for **string** type?

Reply:
@timaschew Thanks for your review.

I used [`vuepress-mergeable`](https://github.com/Shigma/vuepress-mergeable) in this plugin. The `flat` option can flatten all the patterns into a single array. For example, if you apply this plugin multiple times with the following options:

- `['foo']`
- `{ from: 'A', to: 'B' }`
- `['bar', { from: 'C', to: 'D' }]`

The `patterns` will be: `['foo', { from: 'A', to: 'B' }, 'bar', { from: 'C', to: 'D' }]`.

The plugin support both type *string* and *object*, which is similar to [`copy-webpack-plugin`](https://github.com/webpack-contrib/copy-webpack-plugin).

Reply:
@Shigma thanks for explanation, but I think that's not my point.

Let's do some inline comments:
```javascript
// it doesn't matter what type `pattern` is
let patterns = 'test' 

// this ensures that patterns is now an array
if (!Array.isArray(patterns)) patterns = [patterns]

// why you're checking for string type? `pattern` is never a string
pattern = typeof pattern === 'string' ? { from: pattern } : { ...pattern }  
```

Reply:
@timaschew I'm confused by your code.

```ts
let patterns: (string | object)[]

patterns.map((pattern: string | object) => {
  // make sure pattern is an object
  pattern = typeof pattern === 'string'
    ? { from: pattern }
    : { ...pattern }

  // do something
  return pattern
}
```

Reply:
🙈 sorry, somehow I the `map` line was invisible to me